src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,322 @@
+<?php
+/**
+ * Speculative loading functions.
+ *
+ * @package WordPress
+ * @subpackage Speculative Loading
+ * @since 6.8.0
+ */
+
+/**
+ * Returns the speculation rules configuration.
+ *
+ * @since 6.8.0
+ *
+ * @return array<string, string>|null Associative array with 'mode' and 'eagerness' keys, or null if speculative
+ *                                    loading is disabled.
+ */
+function wp_get_speculation_rules_configuration(): ?array {
+	// By default, speculative loading is only enabled for sites with pretty permalinks when no user is logged in.
+	if ( ! is_user_logged_in() && get_option( 'permalink_structure' ) ) {
+		$config = array(
+			'mode'      => 'auto',
+			'eagerness' => 'auto',
+		);
+	} else {
+		$config = null;
+	}
+
+	/**
+	 * Filters the way that speculation rules are configured.
+	 *
+	 * The Speculation Rules API is a web API that allows to automatically prefetch or prerender certain URLs on the
+	 * page, which can lead to near-instant page load times. This is also referred to as speculative loading.
+	 *
+	 * There are two aspects to the configuration:
+	 * * The "mode" (whether to "prefetch" or "prerender" URLs).
+	 * * The "eagerness" (whether to speculatively load URLs in an "eager", "moderate", or "conservative" way).
+	 *
+	 * By default, the speculation rules configuration is decided by WordPress Core ("auto"). This filter can be used
+	 * to force a certain configuration, which could for instance load URLs more or less eagerly.
+	 *
+	 * For logged-in users or for sites that are not configured to use pretty permalinks, the default value is `null`,
+	 * indicating that speculative loading is entirely disabled.
+	 *
+	 * @since 6.8.0
+	 * @see https://developer.chrome.com/docs/web-platform/prerender-pages
+	 *
+	 * @param array<string, string>|null $config Associative array with 'mode' and 'eagerness' keys, or `null`. The
+	 *                                           default value for both of the keys is 'auto'. Other possible values
+	 *                                           for 'mode' are 'prefetch' and 'prerender'. Other possible values for
+	 *                                           'eagerness' are 'eager', 'moderate', and 'conservative'. The value
+	 *                                           `null` is used to disable speculative loading entirely.
+	 */
+	$config = apply_filters( 'wp_speculation_rules_configuration', $config );
```

Comment:
Should we force prevent the feature from being enabled with a filter when a user is logged in? Because of the nature of how flexible WordPress is, allowing this comes with some unknown level of risk, ie. a plugin with no knowledge of sensitive link patterns being defined by another plugin.

Reply:
This relates to what I mentioned in https://core.trac.wordpress.org/ticket/62503#comment:17 (2nd bullet point).

It is disabled by default for logged-in users, and I think that's sufficient. Literally preventing this from being enabled for logged-in users would artificially lock out sites from receiving the behavior. At the end of the day, there are millions of ways to break WordPress and it is the responsibility of the developer of the site to decide which plugins they enable or not. I don't think it would be right to prevent a site owner that knows what they're doing from using a feature that is beneficial to them.

Reply:
Also worth noting, even if a site enables it for logged-in users, it would still not apply to `/wp-admin` due to the exclusions. So that part cannot be circumvented already, which I think is more important.

Reply:
Ah yes! Thanks! I was looking here after the new comments and had forgotten we discussed that on the Trac ticket.